### PR TITLE
sql: error if lacking SELECT privileges on PG source tables

### LINKED ISSUE #19030: pg-cdc: A permission error with one table causes all subsources get borked
### What version of Materialize are you using?

 v0.53.0-dev (c095919b0)

### How did you install Materialize?

Docker image

### What is the issue?

If Mz does not have a permission to perform a `SELECT` on some Postgres table (used during snapshotting), then all subsources of the same Pg source get borked together).

This seems to happen during the snapshot phase -- I assume a global snapshot is used to populate all tables, so if one `SELECT` fails , then the entire source is borked together. 

To reproduce, apply this diff (the pg-cdc-resumption framework is not the best place to have that scenario actually):

```
iff --git a/test/pg-cdc-resumption/configure-postgres.td b/test/pg-cdc-resumption/configure-postgres.td
index d70311441..291855894 100644
--- a/test/pg-cdc-resumption/configure-postgres.td
+++ b/test/pg-cdc-resumption/configure-postgres.td
@@ -11,5 +11,6 @@ $ postgres-execute connection=postgres://postgres:postgres@postgres
 CREATE USER materialize PASSWORD 'materialize';
 ALTER USER materialize REPLICATION;
 GRANT ALL PRIVILEGES ON ALL TABLES IN SCHEMA public TO materialize;
+REVOKE SELECT ON public.ten FROM materialize;
 
 CREATE PUBLICATION mz_source FOR ALL TABLES;
diff --git a/test/pg-cdc-resumption/mzcompose.py b/test/pg-cdc-resumption/mzcompose.py
index ccfacb78d..1b881c4cd 100644
--- a/test/pg-cdc-resumption/mzcompose.py
+++ b/test/pg-cdc-resumption/mzcompose.py
@@ -25,14 +25,7 @@ def workflow_default(c: Composition) -> None:
 
     # TODO: most of these should likely be converted to cluster tests
     for scenario in [
-        disconnect_pg_during_snapshot,
-        disconnect_pg_during_replication,
-        restart_pg_during_snapshot,
-        restart_mz_during_snapshot,
-        restart_pg_during_replication,
-        restart_mz_during_replication,
-        fix_pg_schema_while_mz_restarts,
-        verify_no_snapshot_reingestion,
+        no_disruptions,
     ]:
         print(f"--- Running scenario {scenario.__name__}")
         initialize(c)
@@ -68,6 +61,14 @@ def end(c: Composition) -> None:
     c.run("testdrive", "verify-data.td", "toxiproxy-remove.td")
 
 
+def no_disruptions(c: Composition) -> None:
+    c.run(
+        "testdrive",
+        "delete-rows-t1.td",
+        "delete-rows-t2.td",
+        "alter-table.td",
+    )
+
 def disconnect_pg_during_snapshot(c: Composition) -> None:
     c.run(
         "testdrive",
```

And then run it:

```
cd test/pg-cdc-resumption
./mzcompose down -v ; ./mzcompose run default 
```

it will fail like this:

```
> SELECT COUNT(*) FROM t1;
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 113ms 169ms 253ms 380ms 570ms 854ms 1s 2s 3s 4s 6s 10s 15s 22s 33s 21s
^^^ +++
verify-data.td:10:1: error: executing query failed: db error: ERROR: Source error: u10: db error: ERROR: permission denied for relation ten: ERROR: Source error: u10: db error: ERROR: permission denied for relation ten
     |
   9 | 
  10 | > SELECT COUNT(*) FROM t1;
     | ^
```

A good solution would be to check permissions at purification stage by issuing a `SELECT ... WHERE FALSE LIMIT 0` query and failing the `CREATE SOURCE` command itself. 

### Relevant log output

```source
2023-04-28 10:58:58.854 UTC [201] LOG:  logical decoding found consistent point at 0/A031AC8
2023-04-28 10:58:58.854 UTC [201] DETAIL:  There are no running transactions.
2023-04-28 10:58:58.854 UTC [201] STATEMENT:  CREATE_REPLICATION_SLOT "materialize_f369f2ed8adb41f79577b9027a7500ca" LOGICAL "pgoutput" USE_SNAPSHOT
2023-04-28 10:58:58.891 UTC [201] ERROR:  permission denied for relation ten
2023-04-28 10:58:58.891 UTC [201] STATEMENT:  COPY "public"."ten" TO STDOUT (FORMAT TEXT, DELIMITER '	')
```

### PR DESCRIPTION
**Pull request by sploiselle**
### Motivation

This PR fixes a recognized bug. Fixes #19030

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/postgres-util/src/privileges.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+                lower_bound: 0,
+            },
+        ],
+    );
+
+    let invalid_read_privileges = client
+        .query(
+            "
+            SELECT
+                table_catalog || '.' || table_schema || '.' || table_name AS t
```
Comment (mjibson) [inline src/postgres-util/src/privileges.rs:52]: These surely need escaping and quoting? What if a user has a table name with a double quote or a space?

File: src/postgres-util/src/lib.rs [hunk 2]
```diff
@@ -188,3 +190,8 @@ pub fn make_tls(config: &tokio_postgres::Config) -> Result<MakeTlsConnector, Pos
 
     Ok(tls_connector)
 }
+
+/// Object names surrounded in double quotes are not double-quoted in the system catalog.
+pub(crate) fn sanitize_name_for_catalog_matching(name: &str) -> &str {
+    name.trim_matches('"')
```
Comment (mjibson) [inline src/postgres-util/src/lib.rs:196]: What if the name of the object is a single double quote? trim_matches removes all, not just the first.

File: src/sql/src/pure.rs [hunk 3]
```diff
@@ -407,6 +407,16 @@ pub async fn purify_create_source(
                 });
             }
 
+            // Ensure that we have select permissions on all tables; we have to do this before we
+            // start snapshotting because if we discover we cannot `COPY` from a table while
+            // snapshotting, we break the entire source.
```
Comment (jkosh44) [inline src/sql/src/pure.rs:412]: You probably need USAGE privileges on the schema that the table exists in too. Is it possible that this is ever a view and not a table? In that case you need to make sure that the view owner has enough privileges to execute the query within the view.

I don't know the full context of the issue, but you may also need to check that the user has connect privileges on the database that the table lives in.

File: src/postgres-util/src/privileges.rs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+                &config.get_user().expect("connection specifies user"),
+            ],
+        )
+        .await?
+        .into_iter()
+        .map(|row| row.get("t"))
+        .collect::<Vec<String>>();
+
+    if invalid_read_privileges.is_empty() {
+        Ok(())
```
Comment (jkosh44) [inline src/postgres-util/src/privileges.rs:70]: Can this cause a race condition where the user loses these privileges after checking but before the source is created?

File: src/postgres-util/src/lib.rs (hunk 2 above)
Comment (sploiselle) [reply src/postgres-util/src/lib.rs]: Oh sorry--you caught a bad version of this; it's fixed.

File: src/postgres-util/src/privileges.rs (hunk 4 above)
Comment (sploiselle) [reply src/postgres-util/src/privileges.rs:70]: Yes, but that is always going to be the case unless we jump through a bunch of hoops to not parameterize the query so we can run it inside the transaction that snapshots.

File: src/postgres-util/src/privileges.rs (hunk 1 above)
Comment (sploiselle) [reply src/postgres-util/src/privileges.rs:47]:
```sql
create table """select foo""" (a int);

select distinct table_catalog || '.' || table_schema || '.' || table_name from information_schema.table_privileges where table_name = '"select foo"';
```
```
sean.public."select foo"
```

This generates an error akin to other errors in the process of creating sources, e.g.:

```
CREATE SOURCE "mz_source"
  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
  FOR TABLES (
    "fodo"
  ) with (size = '1');
ERROR:  table fodo not found in source
```
```
CREATE SOURCE "mz_source"
  FROM POSTGRES CONNECTION pg (PUBLICATION 'mz_source')
  FOR TABLES (
    """fodo"""
  ) with (size = '1');
ERROR:  table "fodo" not found in source
materialize=> 
```

File: src/sql/src/pure.rs (hunk 3 above)
Comment (sploiselle) [reply src/sql/src/pure.rs:412]: Ah, great--thanks for catching that. Will add it to the tests.

All of these objects we ingest through publications must be tables.

File: src/postgres-util/src/privileges.rs (hunk 1 above)
Comment (sploiselle) [reply src/postgres-util/src/privileges.rs]: Ah, using `format` with `%I` formatting instructions does all of this nonsense more cleanly and for you. TIL.

**Comment by sploiselle** [comment]
@jkosh44 PTAL at the privilege checks to ensure they make sense; we don't need additional checks for the `CONNECT` privilege because the user cannot connect if they don't have that privilege, which I've added a test for in the new `privileges.td` file.

**Review by philip-stoev — APPROVED** [review]
Thank you for the tests! I can not think of anything else to test about this atm.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Pull request merged**